controller/src/settings/esp.rs

```diff
@@ -93,42 +107,78 @@ impl EspColor {
+                        mid_rgb[0] + (max_rgb[0] - mid_rgb[0]) * t,
+                        mid_rgb[1] + (max_rgb[1] - mid_rgb[1]) * t,
+                        mid_rgb[2] + (max_rgb[2] - mid_rgb[2]) * t,
+                    ]
+                } else {
+                    let t = health * 2.0;
+                    [
+                        min_rgb[0] + (mid_rgb[0] - min_rgb[0]) * t,
+                        min_rgb[1] + (mid_rgb[1] - min_rgb[1]) * t,
+                        min_rgb[2] + (mid_rgb[2] - min_rgb[2]) * t,
+                    ]
+                };
+                [color[0], color[1], color[2], *alpha]
             }
-            Self::HealthBasedRainbow => {
+            Self::HealthBasedRainbow { alpha } => {
                 let sin_value = |offset: f32| {
                     (2.0 * std::f32::consts::PI * health * 0.75 + offset).sin() * 0.5 + 1.0
                 };
                 let r: f32 = sin_value(0.0);
                 let g: f32 = sin_value(2.0 * std::f32::consts::PI / 3.0);
                 let b: f32 = sin_value(4.0 * std::f32::consts::PI / 3.0);
-                [r, g, b, 1.0]
+                [r, g, b, *alpha]
             }
-            Self::DistanceBased => {
+            Self::DistanceBased {
+                near,
+                mid,
+                far,
+                alpha,
+            } => {
                 let max_distance = 80.0;
                 let min_distance = 0.0;
+                // let mid_distance = (max_distance + min_distance) / 2.0;
 
-                let color_near = [1.0, 0.0, 0.0, 0.75];
-                let color_far = [0.0, 1.0, 0.0, 0.75];
+                let color_near = near.as_f32();
+                let color_mid = mid.as_f32();
+                let color_far = far.as_f32();
 
                 let t = (distance - min_distance) / (max_distance - min_distance);
                 let t = t.clamp(0.0, 1.0);
 
-                [
-                    color_near[0] + t * (color_far[0] - color_near[0]),
-                    color_near[1] + t * (color_far[1] - color_near[1]),
-                    color_near[2] + t * (color_far[2] - color_near[2]),
-                    0.75,
-                ]
+                let result = if t < 0.5 {
+                    let t2 = t * 2.0;
+                    [
+                        color_near[0] + t2 * (color_mid[0] - color_near[0]),
+                        color_near[1] + t2 * (color_mid[1] - color_near[1]),
+                        color_near[2] + t2 * (color_mid[2] - color_near[2]),
+                        *alpha,
+                    ]
+                } else {
+                    let t2 = (t - 0.5) * 2.0;
+                    [
+                        color_mid[0] + t2 * (color_far[0] - color_mid[0]),
+                        color_mid[1] + t2 * (color_far[1] - color_mid[1]),
+                        color_mid[2] + t2 * (color_far[2] - color_mid[2]),
+                        *alpha,
+                    ]
+                };
```

Comment:
Make a function from these interpolations

Reply:
should be done

Reply:
didn't test the changes... just assumed it will work, sorry....

Reply:
now